pallets/confidential/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(box_syntax)]
+
+use polymesh_common_utilities::{identity::Trait as IdentityTrait, Context};
+use polymesh_primitives::{AccountId, IdentityId, Ticker};
+use polymesh_primitives_derive::{SliceU8StrongTyped, VecU8StrongTyped};
+
+use pallet_identity as identity;
+
+use bulletproofs::RangeProof;
+use cryptography::asset_proofs::range_proof::{prove_within_range, verify_within_range};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+
+use codec::{Decode, Encode};
+use core::convert::TryFrom;
+use frame_support::{
+    debug, decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult,
+    weights::SimpleDispatchInfo,
+};
+use frame_system::{self as system, ensure_signed};
+use sp_std::prelude::*;
+
+pub mod rng;
+pub use rng::native_rng;
+
+#[derive(Encode, Decode, Clone, Debug, Default, PartialEq, Eq, SliceU8StrongTyped)]
+pub struct RangeProofInitialMessageWrapper(pub [u8; 32]);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped)]
+pub struct RangeProofFinalResponseWrapper(pub Vec<u8>);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct TickerRangeProof {
+    // pub proof: RangeProofWrapper,
+    pub initial_message: RangeProofInitialMessageWrapper,
+    // pub committed_value: CompressedRistrettoWrapper,
+    pub final_response: RangeProofFinalResponseWrapper,
+    pub max_two_exp: u32,
+}
+
+pub trait Trait: frame_system::Trait + IdentityTrait {
+    type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+type Identity<T> = identity::Module<T>;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct ProberTickerKey {
+    prover: IdentityId,
+    ticker: Ticker,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Confidential {
+        /// Number of investor per asset.
+        pub RangeProofs get(fn range_proof): double_map hasher(twox_64_concat) IdentityId, hasher(blake2_128_concat) ProberTickerKey => Option<TickerRangeProof>;
```

Comment:
hmmm, is this line Rust code?

Reply:
It looks like magic because it's inside the `decl_storage!` macro. these macros are a mystery to me too...

Reply:
Yep, `decl_storage!` macro does a lot of magic things. In that one , we are defining a storage which is basically a map with two chained keys. Something like `range_proof [IdentityId] [ProverTickerKey] => Option<TickerRangeProof>`